xdrv_21_wemo.ino

```diff
@@ -0,0 +1,486 @@
+    return String(serial);
+  }
+
+  String WemoUuid(void)
+  {
+    char uuid[29];
+
+    snprintf_P(uuid, sizeof(uuid), PSTR("Socket-1_0-%s"), WemoSerialnumber().c_str());
+    return String(uuid);
+  }
+
+public:
+  WemoSwitch(int deviceId, ESP8266WebServer *webServer)
+  {
+    _deviceId = deviceId;
+    _webServer = webServer;
+
+    AddLog_P2(LOG_LEVEL_DEBUG, PSTR("\nDEVICE #%d LISTENING ON PORT %d"), _deviceId, _localPort);
+  }
+  WemoSwitch(int deviceId, int localPort)
+  {
+    _deviceId = deviceId;
+    _localPort = localPort;
+
+    AddLog_P2(LOG_LEVEL_DEBUG, PSTR("\nDEVICE #%d LISTENING ON PORT %d"), _deviceId, _localPort);
+
+    _webServer = new ESP8266WebServer(_localPort);
+
+    _webServer->begin();
+  }
+
+  ~WemoSwitch()
+  {
+    if (_deviceId > 1)
+    {
+      _webServer->close();
+    }
+  }
+
+  void WemoRespondToMSearch(int echo_type)
+  {
+    AddLog_P2(LOG_LEVEL_DEBUG, PSTR("\nWemoRespondToMSearch DEVICE #%d: %d"), _deviceId, echo_type);
+
+    char message[TOPSZ];
+
+    if (PortUdp.beginPacket(udp_remote_ip, udp_remote_port))
+    {
+      char type[24];
+      if (1 == echo_type)
+      { // type1 echo 1g & dot 2g
+        strcpy_P(type, URN_BELKIN_DEVICE_CAP);
+      }
+      else
+      { // type2 echo 2g (echo, plus, show)
+        strcpy_P(type, UPNP_ROOTDEVICE);
+      }
+      char response[400];
+      snprintf_P(response, sizeof(response), WEMO_MSEARCH, WiFi.localIP().toString().c_str(), _localPort, type, WemoUuid().c_str(), type);
+      PortUdp.write(response);
+      PortUdp.endPacket();
+
+      //AddLog_P2(LOG_LEVEL_DEBUG, PSTR("\nSending packet device %d: %s"), _deviceId, response);
+
+      snprintf_P(message, sizeof(message), PSTR(D_RESPONSE_SENT));
+    }
+    else
+    {
+      snprintf_P(message, sizeof(message), PSTR(D_FAILED_TO_SEND_RESPONSE));
+    }
+    // Do not use AddLog_P2 here (interrupt routine) if syslog or mqttlog is enabled. UDP/TCP will force exception 9
+    PrepLog_P2(LOG_LEVEL_DEBUG, PSTR(D_LOG_UPNP D_WEMO " " D_JSON_TYPE " %d, %s " D_TO " %s:%d"),
+               echo_type, message, udp_remote_ip.toString().c_str(), udp_remote_port);
+  }
+
+  void HandleServerLoop()
+  {
+    if (_webServer != NULL &&  _deviceId > 1)
+    {
+      _webServer->handleClient();
+      delay(1);
+    }
+  }
+
+private:
+  void InternalWSSend(int code, int ctype, const String &content)
+  {
+    char ct[25]; // strlen("application/octet-stream") +1 = Longest Content type string
+    _webServer->send(code, GetTextIndexed(ct, sizeof(ct), ctype, kContentTypes), content);
+  }
+
+  void LogUpnpWithClient(const char *msg)
+  {
+    AddLog_P2(LOG_LEVEL_DEBUG, PSTR(D_LOG_HTTP "%s from %s"), msg, _webServer->client().remoteIP().toString().c_str());
+  }
+
+  void HandleUpnpEvent()
+  {
+    LogUpnpWithClient(PSTR(D_WEMO_BASIC_EVENT));
+
+    char event[500];
+    strlcpy(event, _webServer->arg(0).c_str(), sizeof(event));
```

Comment:
The code was there already but I realize it could be cleaner. First take the length of the buffer and allocate on stack only what's needed. Anyways the UDP packets are capped to `UDP_BUFFER_SIZE` (120)

Reply:
Feel free to apply any kind of optimization!

Reply:
I was thinking of the following:
```
size_t event_len = _webServer->arg(0).length() + 1;
char event[event_len];
strlcpy(event, _webServer->arg(0).c_str(), event_len);
```

Reply:
ok!!